src/java.net.http/share/classes/jdk/internal/net/http/LimitingSubscriber.java

```diff
@@ -0,0 +1,169 @@
+        if (alreadySubscribed) {
+            subscription.cancel();
+        } else {
+            length = 0;
+            downstreamSubscriber.onSubscribe(subscription);
+        }
+    }
+
+    @Override
+    public void onNext(List<ByteBuffer> buffers) {
+
+        // Check arguments
+        requireNonNull(buffers, "buffers");
+        assert Utils.hasRemaining(buffers);
+
+        // See if we can consume the input completely
+        boolean lengthAllocated = allocateLength(buffers);
+        Subscription subscription = subscriptionRef.get();
+        assert subscription != null;
+        if (lengthAllocated) {
+            downstreamSubscriber.onNext(buffers);
+        }
+
+        // See if we can consume the input partially
+        else if (discardExcess) {
+            List<ByteBuffer> retainedBuffers = removeExcess(buffers);
+            if (!retainedBuffers.isEmpty()) {
+                downstreamSubscriber.onNext(retainedBuffers);
+            }
+            downstreamSubscriber.onComplete();
+            subscription.cancel();
+        }
+
+        // Partial consumption is not allowed, trigger failure
+        else {
+            downstreamSubscriber.onError(new IllegalStateException(
+                    "the maximum number of bytes that are allowed to be consumed is exceeded"));
+            subscription.cancel();
+        }
+
+    }
+
+    private boolean allocateLength(List<ByteBuffer> buffers) {
+        long bufferLength = buffers.stream().mapToLong(Buffer::remaining).sum();
+        long nextReceivedByteCount = Math.addExact(length, bufferLength);
+        if (nextReceivedByteCount > capacity) {
+            return false;
+        }
+        length = nextReceivedByteCount;
+        return true;
+    }
+
+    private List<ByteBuffer> removeExcess(List<ByteBuffer> buffers) {
+        List<ByteBuffer> retainedBuffers = new ArrayList<>(buffers.size());
+        long remaining = capacity - length;
+        for (ByteBuffer buffer : buffers) {
+            // No capacity left; stop
+            if (remaining < 1) {
+                break;
+            }
+            // Buffer fits as is; keep it
+            else if (buffer.remaining() <= remaining) {
+                retainedBuffers.add(buffer);
+                remaining -= buffer.remaining();
+            }
+            // There is capacity, but the buffer doesn't fit; truncate and keep it
+            else {
+                buffer.limit(Math.toIntExact(Math.addExact(buffer.position(), remaining)));
+                retainedBuffers.add(buffer);
+                remaining = 0;
+            }
+        }
+        return retainedBuffers;
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        requireNonNull(throwable, "throwable");
+        downstreamSubscriber.onError(throwable);
+    }
+
+    @Override
+    public void onComplete() {
+        downstreamSubscriber.onComplete();
+    }
```

Comment:
As hinted earlier you should not call onError again if you already called onError. The publisher should not call onError twice, but if you have already called onError from onNext on the downstream subscriber, you should not call it again (and neither should you call onComplete), if the upstream publisher calls onError/onComplete on the limiting subscriber.

Reply:
Added state and relevant checks in a4fd1bab9cb50eca5f2040ca5e27798f19d5abfa.